solution1: add solution for exercise 1

kopiczko opened the pull request:
# WIP

kopiczko commented:
@rossf7 this is still WIP, but you can check my approach. Any criticism welcome.

rossf7 commented on solution1/crd.yaml:
```diff
@@ -0,0 +1,26 @@
+# 
+# NOTE: This file is not used in the solution this serves an example how the yaml
+# file looks like. Content of this file is copied crdJson variable in JSON
+# format.
```
Having the yaml file is nice <3 Nitpick but I'd rewrite the comment like this.

This file is not used in the solution. It is an example of what the YAML file looks like. Content in this file is copied from the crdJson JSON variable.

kopiczko replied:
Done.

rossf7 commented on solution1/solution1.go:
```diff
@@ -0,0 +1,180 @@
+func parseFlags() Config {
+	var config Config
+
+	var homedir string
```
Nitpick but I prefer homeDir.

kopiczko replied:
Done.

rossf7 commented on solution1/solution1.go:
```diff
@@ -0,0 +1,180 @@
+		log.Printf("creating custom resource")
+
+		res, err := k8sClient.Post(
+			config.K8sServer+"/apis/apiextensions.k8s.io/v1beta1/customresourcedefinitions",
```
A constant here would be good. At the end we can show how operatorkit helps with paths.

kopiczko replied:
I agree, but I don't want ppl to jump around, I'd like them to see immediately the path. We shouldn't consider it production grade code. WDYT?

rossf7 replied:
Good point, no need for a constant.

rossf7 commented on solution1/solution1.go:
```diff
@@ -0,0 +1,180 @@
+		case http.StatusConflict:
+			m := make(map[string]interface{})
+			err := json.NewDecoder(res.Body).Decode(&m)
+			return fmt.Errorf("creating custom resource: decoding body=%s", readerToString(res.Body))
```
Can the log say it's creating the resource definition?  I can see newbies getting confused between CRDs and CROs at the start.

kopiczko replied:
I think it is ok to call it `custom resource` and `custom object`. (1) it's interchangeable with TPR stuff. (2) Below you have a log for waiting for custom resource to be ready, which is something different than waiting for custom resource definition to be ready. Custom resource is something Kubernetes creates as a result of custom resource definition object reconciliation, and we should make that clear. Do you agree?

rossf7 replied:
Yes, I like custom resource and custom object. Nice that it's interchangeable with TPRs.

rossf7 commented on solution1/solution1.go:
```diff
@@ -0,0 +1,180 @@
+	flag.StringVar(&config.K8sCrtFile, "kubernetes.crt", path.Join(homedir, ".minikube/apiserver.crt"), "Kubernetes certificate file path.")
+	flag.StringVar(&config.K8sKeyFile, "kubernetes.key", path.Join(homedir, ".minikube/apiserver.key"), "Kubernetes key file path.")
+	flag.StringVar(&config.K8sCAFile, "kubernetes.ca", path.Join(homedir, ".minikube/ca.crt"), "Kubernetes CA file path.")
+	flag.Parse()
```
I like how this uses the minikube config.

kopiczko commented:
@rossf7 I think the solution for ex1 is ready. Last bit is to implement `MySQLOps`.
I haven't addressed your comments yet, but I'll do! Thanks for leaving them!

kopiczko commented on solution1/solution1.go:
```diff
@@ -0,0 +1,428 @@
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
```
Note to self: remove comments below. This was my scratch area :D

kopiczko replied:
Done.

rossf7 commented on solution1/solution1.go:
```diff
@@ -0,0 +1,404 @@
+	// Start reconciliation loop.
+	reconciliationCnt := 1
+	reconciliationInterval := time.Second * 1
+	for ; ; reconciliationCnt++ {
```
Nitpick but the comment here could say it's an infinite loop and that its intended.

kopiczko commented on postgresqlops/postgresqlops.go:
```diff
@@ -109,7 +117,7 @@ func (p *PostgreSQLOps) DeleteDatabase(name string) error {
 	dbs := []Database{}
 
-	rows, err := p.db.Query("SELECT pg_database.datname, pg_user.usename FROM pg_database, pg_user WHERE pg_database.datdba = pg_user.usesysid")
+	rows, err := p.db.Query("SELECT pg_database.datname, pg_user.usename FROM pg_database, pg_user WHERE pg_database.datdba = pg_user.usesysid AND pg_database.datname NOT IN ('postgres', 'template0', 'template1')")
```
I added some table filtering.

rossf7 replied:
Good catch! That would cause problems if they were returned.

kopiczko commented on postgresqlops/postgresqlops.go:
```diff
@@ -42,6 +45,11 @@ func New(config Config) (*PostgreSQLOps, error) {
+// Close relases all PosgreSQLOps resources.
+func (p *PostgreSQLOps) Close() error {
+	return p.db.Close()
+}
```
Added close function.

kopiczko commented on postgresqlops/postgresqlops.go:
```diff
@@ -28,7 +31,7 @@ type PostgreSQLOps struct {
 func New(config Config) (*PostgreSQLOps, error) {
 	// Postgres user and password are hardcoded and match the resources in postgres.yaml.
-	psqlInfo := fmt.Sprintf("host=%s port=%d user=%s password=%s sslmode=disable", config.Host, config.Port, "postgres", "operator-workshop")
+	psqlInfo := fmt.Sprintf("host=%s port=%d user=%s password=%s sslmode=disable", config.Host, config.Port, config.User, config.Password)
```
User/passwd moved from here.

kopiczko commented on solution1/solution1.go:
```diff
@@ -0,0 +1,360 @@
+	"time"
+
+	"github.com/giantswarm/operator-workshop/postgresqlops"
+)
```
I split into multiple files, so this has only interesting code, and not e.g. os signal handling.

rossf7 commented on solution1/crd.yaml:
```diff
@@ -0,0 +1,26 @@
+kind: CustomResourceDefinition
+metadata:
+  # name must match the spec fields below, and be in the form: <plural>.<group>
+  name: mysqlconfigs.containerconf.de
```
Can you update the CRD example for PostgreSQL.
